// proto/schema/componnet.proto
syntax = "proto3";

package componnet;

message GridAttribute {
    bool deleted = 1;
    bool activate = 2;
    int32 type = 3;
    int32 level = 4;
    int32 global_id = 5;
    double elevation = 6;
    optional double min_x = 7;
    optional double min_y = 8;
    optional double max_x = 9;
    optional double max_y = 10;
    optional int32 local_id = 11;
}

message SubdivideRule {
    repeated int32 rule = 1;
}

message InitParams {
    string redis_host = 1;
    int32 redis_port = 2;
    int32 epsg = 3;
    repeated double bounds = 4; // [min_x, min_y, max_x, max_y]
    repeated float first_size = 5; // [width, height]
    repeated SubdivideRule subdivide_rules = 6;
}

message PeerGridInfos {
    int32 level = 1;
    repeated int32 global_ids = 2;
}

message GridInfos {
    repeated int32 levels = 1;
    repeated int32 global_ids = 2;
}

message LocalIds {
    repeated int32 local_ids = 1;
}

message GridBBoxes {
    repeated double min_xs = 1;
    repeated double min_ys = 2;
    repeated double max_xs = 3;
    repeated double max_ys = 4;
}

message GridAttributes {
    repeated GridAttribute attributes = 1;
}

message GridKeys {
    repeated string grid_keys = 1;
}

message ComponentRequest {
    string crm_address = 1;
    oneof args {
        PeerGridInfos peerGridInfos = 2;
    }
}

service ComponentService {
    rpc Execute (ComponentRequest) returns (GridAttributes) {}
}
